Wip - live mode

bentiss opened the pull request:
Still a WIP as it doesn't work and need some tuning for the created uhid node.

Note that `tuhi-live.py` requires root but drops the privileges ASAP, and only keep the root access for opening the uhid node. This is required if we want to be able to connect to the user session bus.

The fd passing doesn't seem to be working now, not sure why.

Still missing the README changes, as the API is not final.

Talking about the API, I wonder if we couldn't have a mode where we provide the data over dbus (so tuhi-kete can use it), and an other where we provide the uhid fd for direct kernel injection.

Related to #43.

Things to fix before merging:
- [x] some feedback when the device is connected and the live mode has started - **kete now emulates a uhid /dev/node**
- [ ] some hardening when the device is already connected or so
- [ ] fix the physical dimensions too - **see #84**
- [ ] far too many points of errors if anything doesn't go well
- [ ] kill the second battery node exported by `wacom.ko` - **see #85**
- [ ] use a proper pid, not `0x0001` - **see #86**
- [ ] make gnome-control-center detect the tablet (libwacom?) - **see #87**

whot commented on tools/tuhi-kete.py:
```diff
@@ -541,6 +549,38 @@ def json_to_svg(self, js, filename):
         svg.save()
 
 
+class LiveChanger(Worker):
```
haha, that class name :)

bentiss replied:
I knew this would get you a good chuckle. I had one too and then decided I couldn't change the name anymore, too much fun :)

whot commented on tools/tuhi-kete.py:
```diff
@@ -1003,6 +1043,67 @@ def do_info(self, args):
+
+    def do_enable_live(self, args):
+        desc = '''
+        Enable or disable live mode on a particular device'''
```
strange linebreaks here

whot commented on tuhi/wacom.py:
```diff
@@ -173,7 +210,9 @@ def _on_pen_data_changed(self, name, value):
             while data:
                 if bytes(data) == b'\xff\xff\xff\xff\xff\xff':
                     logger.info(f'Pen left proximity')
+                    self.pen_logger.info(f'E: {int(self._timestamp/1000)}.{self._timestamp % 1000:03} 8 01 00 00 00 00 00 00 00')
```
that hex stuff should be a #define thingy

bentiss replied:
Yeah, this is just some debugging that are doomed to be removed in the final PR

whot commented on tuhi/wacom.py:
```diff
@@ -173,7 +210,9 @@ def _on_pen_data_changed(self, name, value):
                     logger.info(f'Pen left proximity')
+                    self.pen_logger.info(f'E: {int(self._timestamp/1000)}.{self._timestamp % 1000:03} 8 01 00 00 00 00 00 00 00')
                 else:
+                    self.pen_logger.info(f'E: {int(self._timestamp/1000)}.{self._timestamp % 1000:03} 8 01 01 {b2hex(bytes(data[:6]))}')
```
same here

whot commented on tuhi/wacom.py:
```diff
@@ -337,6 +376,11 @@ def ec_command(self):
+        self.pen_logger.info(f'R: {len(rdesc)} {" ".join([f"{r:02x}" for r in rdesc])}')
+        self.pen_logger.info(f'N: {self.name}')
+        self.pen_logger.info(f'I: 5 056a 0001')
+        self.pen_logger.info(f'Starting wacom live mode')
```
isn't that line going to mess up your parsing of the output otherwise? Plus, the raw data should be debug, no()

whot commented on tuhi/base.py:
```diff
@@ -192,7 +192,7 @@ def _on_live_updated(self, dbus_device, pspec):
         else:
             if self._wacom_device is not None:
-                self._wacom_device.start(False, False)
+                self._wacom_device.start(False, False, -1)
```
this is getting to the point where it'd be better to use kwargs

bentiss replied:
yep. I also think we should rethink the way we start the wacom device. But it was just too much for now, I am trying to make it working somehow for now.

whot commented on tuhi/dbusserver.py:
```diff
@@ -167,6 +167,7 @@ def __init__(self, device, connection):
         self._listening = False
         self._listening_client = None
         self._live = False
+        self._uhid = None
```
can we name this `_uhid_fd` please?

bentiss replied:
sure, that's why it's a wip

whot commented on tuhi/dbusserver.py:
```diff
@@ -391,6 +396,7 @@ def _start_live(self, connection, sender, args):
         logger.debug(f'Live mode started on {self.name} for {sender}')
 
         self.live = True
+        self._uhid = args[0]
```
In regards to this not working, did you print the type of `args[0]` here? maybe it's still wrapped in something?

bentiss replied:
could be, I'll have a look

whot commented on tools/tuhi-live.py:
```diff
@@ -0,0 +1,203 @@
+            return 0
+        else:
+            fd = os.open('/dev/uhid', os.O_RDWR)
+            reduction.send_handle(conn_out, fd, pid)
```
one example that I saw was using `fd.fileno()` here. maybe that makes a difference?

bentiss replied:
It shouldn't. It depends if you are using `File` objects or if you are using, like here plain fds. The only difference is that with plain fds, you have to use `os.read/write` instead of `f.read/write`

whot reviewed (commented):
Ok, way too much for my brain today, but I'll leave you with these comments anyway... sorry

whot reviewed (commented):
christ, github is not happy with multiple tabs...

whot reviewed (commented):
github not happy with multiple tabs, sorry

whot reviewed (commented):
sigh

whot commented:
ok, not sure what exactly happened here, but github lost half my comments and now they show as outdated....

bentiss commented:
I have pushed a new branch with lots of improvements: I can now use the slate as a regular input node :)

~Things to fix before merging:~ 

EDIT: moved into first post

bentiss committed:
live: add a script to run as root that enable live mode

we want to be able to open uhid, so we need root access.
Root access is bad, so we fork the process, and then immediately drop
the privileges to talk to the dbus dameon (or to create one).

bentiss committed:
live-mode: properly pack the fds over DBus

Keeping this one as a separate commit, so we can use it as a future
reference.

bentiss committed:
kete: parse incoming uhid data

bentiss committed:
tuhi: hook up live mode

Currently just enable/disable live mode.
The int parameter is doomed to be an opened fd to the uhid node

bentiss committed:
wacom: insert proper ranges for X and Y in the uhid device

Also fix the pressure range to be 2047

bentiss committed:
kete: rename the script from tuhi-kete.py to kete.py

Python doesn't really like minus signs in file names, especially
when you start importing the modules. Rename kete from tuhi-kete.py
to kete.py so it makes it easier to have other scripts in the tools
folder to borrow code from kete.

bentiss committed:
kete: export the connected state of the bluez device to the tuhiketedevice

We need this in the live mode as the connection will be longer than
the usual retrieval of the data.

bentiss committed:
wacom: create a uhid device on live mode

uhid code taken from https://github.com/bentiss/hid-tools and stripped out
from the not required parts here

bentiss committed:
kete: add live debugging mode

Just start/stop live mode, the debug happens in the tuhi process.

That's enough until we figure out the DBus API if we need one

bentiss commented:
Removing the `wip - do not merge tag` as I am tired rebasing this on top of master. It takes forever each time.

Now, kete has some feedback from tuhi when enabling the live mode by emulating a uhid device. This allows to keep the DBus API simple (you have to provide a valid fd), and get some data out of kete.
